src/clj_puppetdb/http.clj

```diff
@@ -112,26 +153,21 @@
   automatically and added to the path."
   ([client :- Client ^String path params]
     {:pre (map? params)}
-    (let [query (if (empty? params)
-                  path
-                  (str path "?" (map->query params)))]
-      #_(println "GET:" query)                                 ;; uncomment this to watch queries
-      (let [{:keys [host opts]} client
-            vcr-dir (:vcr-dir opts)
-            opts (dissoc opts :vcr-dir)
-            response (-> (str host query)
-                         (http-get (assoc opts :as :stream) vcr-dir)
-                         (catching-exceptions {:kind :puppetdb-connection-error}))]
-        (if-not (= 200 (:status response))
-          (throw (ex-info nil {:kind   :puppetdb-query-error
-                               :url    (str host path)
-                               :msg    (slurp (util/make-response-reader response))
-                               :status (:status response)
-                               :params params})))
-        (let [data (-> response
-                       util/make-response-reader
-                       decode-stream-catching-parse-exceptions)
-              headers (:headers response)]
-          [data headers]))))
+    (let [query-info (-> (client)                           ; get client info
+                         (assoc :endpoint path)
+                         (assoc :params   params))
+          response (-> (client client path params)
```

Comment:
Unfortunately this then prohibits the overriding from vcr enabled client.
The vcr enabled client calls the standard client while passing itself in the ```this``` argument thus making the standard client call the vcr client's version of the of the different arity overloaded variants of the client function.
The sequence of calls is then as follows:
1. the GET method calls the vcr client's [this path params] arity overloaded variant passing the vcr client in the ```this``` parameter
2. the vcr client's [this path params] sorts the known nested structures in params and calls the standard client's [this path params] arity overloaded variant passing on the ```this``` parameter unchanged (i.e. pointing to the vcr client)
3. the standard client's [this path params] forms the PDB query URL by converting the params into a query string and concatenating it to the ```host``` and ```path``` and uses the ```this``` parameter to call the [query] arity overloaded variant - which means (given that ```this``` points to the vcr client) that it calls the vcr client's version of that arity overloaded variant in this case
4. vcr client's [query] arity overloaded variant first tries to look up a response for the query in the ```vcr-dir``` and when that fails it calls the standard client's version of the [query] arity overloaded variant to actually query the PDB after which the vcr client's [query] arity overloaded variant saves the response for future use